src/rust/options/src/pants_ng_flags.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2025 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use crate::fromfile::FromfileExpander;
+use crate::id::NameTransform;
+use crate::parse::{ParseError, Parseable};
+use crate::{DictEdit, ListEdit, OptionsSource};
+use crate::{OptionId, Scope};
+use std::any::Any;
+use std::collections::HashMap;
+
+// This properly belongs in pants_ng/options/flags.rs, but it depends on a lot of non-public
+// code from this crate, so we keep it here for now rather than open up this crate.
+pub struct PantsNgFlagsReader {
+    // Map from scope to map of name => vals. Multiple vals means that the same
+    // flag name was specified multiple times.
+    flags: HashMap<Scope, HashMap<String, Vec<Option<String>>>>,
+    fromfile_expander: FromfileExpander,
+}
+
+impl PantsNgFlagsReader {
+    pub fn new(
+        flags: HashMap<Scope, HashMap<String, Vec<Option<String>>>>,
+        fromfile_expander: FromfileExpander,
+    ) -> Self {
+        Self {
+            flags,
+            fromfile_expander,
+        }
+    }
+
+    fn get_vals(&self, id: &OptionId) -> Option<&Vec<Option<String>>> {
+        // TODO: memoize name_underscored() onto OptionId?
+        self.flags
+            .get(&id.scope)
+            .and_then(|hm| hm.get(&id.name_underscored()))
+    }
+
+    fn to_bool(&self, val: &Option<String>) -> Result<Option<bool>, ParseError> {
+        // An arg can represent a bool either by having an explicit value parseable as a bool,
+        // or by having no value (in which case it represents true).
+        match val {
+            Some(value) => match self.fromfile_expander.expand(value.to_string())? {
+                Some(s) => bool::parse(&s).map(Some),
+                _ => Ok(None),
+            },
+            None => Ok(Some(true)),
+        }
+    }
+
+    fn get_list<T: Parseable>(&self, id: &OptionId) -> Result<Option<Vec<ListEdit<T>>>, String> {
+        let mut edits = vec![];
+        if let Some(vals) = self.get_vals(id) {
+            for opt_val in vals {
+                if let Some(val) = opt_val {
```

Comment:
Feels like we could use some sweet sweet let-else chains from 2024 here

Reply:
Looked at this, not sure there is a much simpler way to express it.